### ISSUE TITLE
Startup fails when a directory is introduced in grpc/protos

### ISSUE THREAD

**Issue by ankitanallana-livongo** [testinggospels/camouflage#76]
**Describe the bug**
Camouflage startup fails when I add a directory to the `grpc/protos` folder.
The error I see is:
```
2021-07-27 16:38:14 warn: [84661] Worker Stopped Tue Jul 27 2021 16:38:14 GMT-0700 (Pacific Daylight Time)
2021-07-27 16:38:15 info: CORS enabled for http://localhost:3000/, http://localhost:3001/, http://localhost:5000/
2021-07-27 16:38:15 info: [84664] Worker started
2021-07-27 16:38:15 info: Handlebar helpers registration started
2021-07-27 16:38:15 warn: Code Injection is enabled.
2021-07-27 16:38:15 error: Loading custom handlebar helpers from ./custom_handlebar.json failed. File not found.
2021-07-27 16:38:15 info: Handlebar helpers registration completed
/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/root.js:104
            throw err;
            ^

Error: EISDIR: illegal operation on a directory, read
    at Object.readSync (node:fs:720:3)
    at tryReadSync (node:fs:430:20)
    at Object.readFileSync (node:fs:476:19)
    at fetch (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/root.js:172:34)
    at Root.load (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/root.js:206:13)
    at Root.loadSync (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/root.js:247:17)
    at Object.loadProtosWithOptionsSync (/usr/local/lib/node_modules/camouflage-server/node_modules/@grpc/proto-loader/build/src/util.js:66:29)
    at Object.loadSync (/usr/local/lib/node_modules/camouflage-server/node_modules/@grpc/proto-loader/build/src/index.js:191:31)
    at /usr/local/lib/node_modules/camouflage-server/dist/protocols/index.js:106:46
    at Array.forEach (<anonymous>) {
  errno: -21,
  syscall: 'read',
  code: 'EISDIR'
}
```

**To Reproduce**
1. Create any directory under `${CAMOUFLAGE_DIR}/grpc/protos`
2. Start up camouflage `%  camouflage --config config.yml`

**Expected behavior**
Startup without any issues

**Desktop (please complete the following information):**
 - OS: MacOS Catalina
 - Browser - N/A - using `grpcurl` for grpc requests
 - Version - `0.5.2`

**Additional context**
Tried out a sample grpc mock (the `todoService.readTodos`) which worked as expected. Tried to add a grpc mock for our use case which contains a couple of protos organised in directories. But startup fails when a directory is introduced.

**Comment by shubhendumadhukar** [comment]
@ankitanallana-livongo : Could you please elaborate a bit more on the issue?

> Tried to add a grpc mock for our use case which contains a couple of protos organised in directories. But startup fails when a directory is introduced.

This is not really required. You could simply place all your proto files in the grpc/protos directory (no need to create additional directories). A necessary folder structure is only needed when you create mock files.

**Comment by ankitanallana-livongo** [comment]
I see. I tried it that way with `grpcurl`:

```
grpcurl -vv  -plaintext \
       -import-path /Users/ankita.nallana/camouflage/grpc/protos \
       -proto sso_api.proto localhost:4312 com.livongo.protobuf.grpc.external.enrollment.Sso.RouteExternalUser

Failed to process proto source files.: could not parse given files: sso_api.proto:5:8: open /Users/ankita.nallana/camouflage/grpc/protos/livongo/protobuf/grpc/external/enrollment/sso/route_external_users.proto: no such file or directory
```
Both `sso_api.proto` and `route_external_users.proto` are in `grpc/protos` folder but I think the imports(`import "livongo/protobuf/grpc/external/enrollment/sso/route_external_users.proto";` in `sso_api.proto`) ensure a directory/package structure.

**Comment by shubhendumadhukar** [comment]
That gives more clarity on the issue. Thank you.

Let me look into it.

**Comment by shubhendumadhukar** [comment]
@ankitanallana-livongo : The error above seems to be coming from grpcurl due to the imports as you rightly mentioned. Though it's not a Camouflage error, it does require a certain folder structure. I have made necessary changes allowing Camouflage to search for proto files in grpc/protos recursively, so you can now add sub directories under grpc/protos. However I don't think that's the reason, you are getting this error.

To confirm my assumption, I'd like to run a triage. Would you be able to change the `-import-path` to a directory other than `grpc/protos`?
- Camouflage set up would remain unaffected. i.e. keep all relevant proto files under `grpc/protos`
- Create another new directory for your proto files elsewhere with necessary folder structure as required by your imports.
- Specific the new directory in the `-import-path` property while running the grpcurl.

To summarize, we want to specify separate locations for proto files, for client and server.

Let me know if that makes sense?

**Comment by ankitanallana-livongo** [comment]
@shubhendumadhukar is version `0.5.3` available via `npm` yet? I see the following error for `0.5.3` - `npm install camouflag@latest` gets me version 0.5.2:
```
npm install camouflage@0.5.3
npm ERR! code ETARGET
npm ERR! notarget No matching version found for camouflage@0.5.3.
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.
```

**Comment by shubhendumadhukar** [comment]
@ankitanallana-livongo : I have not published a new release yet. As I mentioned in the last comment, the issue you are facing, might not be because of Camouflage.

Could you please let me know the outcome of the triage  I mentioned in previous comment?

>To confirm my assumption, I'd like to run a triage. Would you be able to change the -import-path to a directory other than grpc/protos?

> Camouflage set up would remain unaffected. i.e. keep all relevant proto files under grpc/protos

> Create another new directory for your proto files elsewhere with necessary folder structure as required by your imports.

> Specify the new directory in the -import-path property while running the grpcurl.

>To summarize, we want to specify separate locations for proto files, for client and server.

**Comment by ankitanallana-livongo** [comment]
@shubhendumadhukar 

I did try what you mentioned - I moved all relevant proto files to under `grpc/protos` and readied my `grpcurl` point to another directory with the necessary structure. 

But, startup failed yet again because Enums/Types could not be resolved despite them being present in the `grpc/protos` folder. Here's what I see:

```
2021-07-28 15:26:06 info: Handlebar helpers registration completed
/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/namespace.js:411
        throw Error("no such Type or Enum '" + path + "' in " + this);
        ^

Error: no such Type or Enum 'SemanticVersion' in Type .com.livongo.protobuf.common.types.ProtocolVersion
    at Type.lookupTypeOrEnum (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/namespace.js:411:15)
    at Field.resolve (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/field.js:268:94)
    at Type.set (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/type.js:177:38)
    at Type.get (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/type.js:155:45)
    at Field.resolve (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/field.js:317:21)
    at Type.resolveAll (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/type.js:304:21)
    at Namespace.resolveAll (/usr/local/lib/node_modules/camouflage-server/node_modules/protobufjs/src/namespace.js:308:25)

```
I double-checked - proto file for `SemanticVersion` is present in `grpc/protos` but I'm not sure now ...

**Comment by shubhendumadhukar** [comment]
Thank you.

This might require some further debugging. I'll publish 0.5.3 in couple of hours.

If you'd like to test the changes meanwhile you could build the package locally from `develop` branch. Instructions to build are in README

**Comment by shubhendumadhukar** [comment]
The fix would only work if the imported proto file does not have any service/method defined, otherwise it tries to register same method twice. Looking into it.

**Comment by shubhendumadhukar** [comment]
@ankitanallana-livongo : This should work now. Could you please validate with 0.5.5? If not, please reopen this issue.

**Comment by ankitanallana-livongo** [comment]
@shubhendumadhukar I no longer see the `EISDIR` error while `Camouflage` starts up :)
